### PR TITLE
feat(rtl,client-gen): Support event subscription for GTest via stream

### LINKED ISSUE #299: rtl, client-gen: Support event subscription for GTest
### Problem to Solve

At the moment there is no option to subscribe to events emitted by program

### Possible Solution

 Introduce some trait for receiving events as bytes and implement it for GTestRemoting. Generate some event listener on a per service basis which allows to receive a typed event

### Notes

_No response_

### PR DESCRIPTION
**Pull request by vobradovich**
Resolves #299

### CODE REVIEW

File: rtl/src/gtest/calls.rs [hunk 1]
```diff
@@ -135,3 +155,40 @@ impl Remoting<GTestArgs> for GTestRemoting {
...
+impl Listener<Vec<u8>> for GTestRemoting {
+    async fn listen(&mut self) -> Result<impl Stream<Item = (ActorId, Vec<u8>)>> {
+        let listener = Rc::new(GTestEventListener::default());
+        self.listeners.borrow_mut().push(Rc::downgrade(&listener));
+        Ok(GTestEventStream(Pin::new(listener)))
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct GTestEventListener {
```
Comment (DennisInSky) [inline rtl/src/gtest/calls.rs:168]: Doesn't seem needed to be public

File: rtl/src/gtest/calls.rs [hunk 2]
```diff
@@ -135,3 +155,40 @@ impl Remoting<GTestArgs> for GTestRemoting {
...
+    events: RefCell<Vec<(ActorId, MessageId, Vec<u8>)>>,
+}
+
+impl GTestEventListener {
+    pub fn events(&self) -> impl Deref<Target = Vec<(ActorId, MessageId, Vec<u8>)>> + '_ {
+        self.events.borrow()
+    }
+}
+
+pub struct GTestEventStream(Pin<Rc<GTestEventListener>>);
```
Comment (DennisInSky) [inline rtl/src/gtest/calls.rs:178]: Doesn't seem need to be public

File: rtl/src/gtest/calls.rs [hunk 3]
```diff
@@ -73,6 +78,20 @@ impl GTestRemoting {
...
+        let events: Vec<(ActorId, MessageId, Vec<u8>)> = run_result
+            .log()
+            .iter()
+            .filter(|entry| entry.destination() == ActorId::zero())
+            .map(|entry| (entry.source(), entry.id(), entry.payload().to_vec()))
+            .collect();
+        for listener in self.listeners.borrow_mut().iter_mut() {
+            if let Some(listener) = listener.upgrade() {
+                listener.events.borrow_mut().append(&mut events.clone());
+            }
```
Comment (DennisInSky) [inline rtl/src/gtest/calls.rs:91]: Do you want to drop dead weak refs from the vec?

File: client-gen/tests/snapshots/generator__events_works.snap [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                if !payload.starts_with(SERVICE_ROUTE) {
+                    Err(sails_rtl::errors::RtlError::EventPrefixMismatches)?;
+                }
+                let event_bytes = &payload[0..];
+                for (idx, name) in EVENT_NAMES.iter().enumerate() {
+                    if event_bytes.starts_with(name) {
+                        let idx = idx as u8;
+                        let bytes = [&[idx], &event_bytes[name.len()..]].concat();
+                        let mut event_bytes = &bytes[..];
+                        return Ok(ServiceWithEventsEvents::decode(&mut event_bytes)?);
```
Comment (DennisInSky) [inline client-gen/tests/snapshots/generator__events_works.snap:106]: Won't it be enough?
```suggestion
                        return ServiceWithEventsEvents::decode(&mut event_bytes);
```

File: client-gen/tests/snapshots/generator__events_works.snap [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+    #[allow(dead_code)]
+    #[cfg(not(target_arch = "wasm32"))]
+    pub mod events {
+        use super::*;
+        use sails_rtl::events::{Listener, RemotingListener};
+        #[derive(PartialEq, Debug, Encode, Decode)]
+        #[codec(crate = sails_rtl::scale_codec)]
+        pub enum ServiceWithEventsEvents {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply client-gen/tests/snapshots/generator__events_works.snap:81]: much better with the `events` mod - separate it from calls, local `use`.
The `events` mod is not generated at all if the service has no events

File: examples/demo/wasm/tests/counter_gtest.rs [hunk 6]
```diff
@@ -28,3 +29,71 @@ async fn counter_works() {
 
     assert_eq!(result, 52);
 }
+
+#[tokio::test]
+async fn counter_events() {
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply examples/demo/wasm/tests/counter_gtest.rs:34]: different tests for different things, previous tests are for calls

File: examples/demo/wasm/tests/counter_gtest.rs [hunk 7]
```diff
@@ -28,3 +29,71 @@ async fn counter_works() {
...
+    println!("{:?}", event);
+    assert_eq!(
+        CounterEvents::Subtracted(1),
+        CounterEvents::decode_event(event.1).unwrap()
+    );
+
+    let event = listener.next().await.unwrap();
+    assert_eq!((program_id, CounterEvents::Added(2)), event);
+    let event = listener.next().await.unwrap();
+    assert_eq!((program_id, CounterEvents::Subtracted(1)), event);
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply examples/demo/wasm/tests/counter_gtest.rs:98]: em?
```
    assert_eq!((program_id, CounterEvents::Added(2)), event);
```

File: rtl/src/gtest/calls.rs (hunk 3 above)
Comment (vobradovich) [reply rtl/src/gtest/calls.rs:91]: non neccesary, couse it all for tests only

File: client-gen/tests/snapshots/generator__events_works.snap (hunk 4 above)
Comment (vobradovich) [reply client-gen/tests/snapshots/generator__events_works.snap:106]: No, to convert the error

File: examples/demo/wasm/tests/counter_gtest.rs [hunk 8]
```diff
@@ -28,3 +29,71 @@ async fn counter_works() {
...
     assert_eq!(result, 52);
 }
+
+#[tokio::test]
+async fn counter_events() {
+    let fixture = Fixture::new(fixture::ADMIN_ID);
+
+    // Low level remoting listener
+    let mut space = fixture.program_space().clone();
+    let mut remoting_listener = space.listen().await.unwrap();
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply examples/demo/wasm/tests/counter_gtest.rs]: hm, listen take &mut self, so space must outlive remoting_stream

File: rtl/src/gtest/calls.rs (hunk 3 above)
Comment (vobradovich) [reply rtl/src/gtest/calls.rs:91]: added containers.retain berfore append new events

File: examples/demo/wasm/tests/counter_gtest.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (vobradovich) [reply examples/demo/wasm/tests/counter_gtest.rs]: added for low level

File: examples/demo/wasm/tests/counter_gtest.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (DennisInSky) [reply examples/demo/wasm/tests/counter_gtest.rs]: These tests are not for testing generated infra, but for testing contract behaviour, i.e. it does what needs to do. The infra stuff is tested here implicitly. Explicit testing of the latter requires a different test suite residing next to the responsible code.

File: client-gen/tests/snapshots/generator__events_works.snap (hunk 4 above)
Comment (DennisInSky) [reply client-gen/tests/snapshots/generator__events_works.snap:106]: `into()` ?

File: rtl/src/gtest/calls.rs [hunk 9]
```diff
@@ -1,13 +1,18 @@
...
     errors::{Result, RtlError},
+    events::Listener,
     prelude::*,
     rc::Rc,
 };
-use core::future::Future;
+use core::{cell::RefCell, future::Future, pin::Pin, task::Poll};
+use futures::Stream;
 use gear_core_errors::{ReplyCode, SuccessReplyReason};
+use gstd::rc::Weak;
```
Comment (DennisInSky) [inline rtl/src/gtest/calls.rs:11]: Prefer using crate::rc::Weak

Comment (vobradovich) [reply rtl/src/gtest/calls.rs:11]: are no longer needed)

**Review by DennisInSky — APPROVED** [review]
Nice one

**Pull request merged**